### PR TITLE
CP-15358: support external ip address setting

### PR DESCRIPTION
**Pull request by Chunjie**
extend vif-param-set to set following xenstore keys, then windows guest agent reads these keys to write static ip setting to windows guest os

xenserver
    device = ""
         vif = ""
             0 = ""
                  static-ip-setting = ""
                      mac = "ae:0e:ad:e7:a5:6e"
                      error-code = "0"
                      error-msg = ""
                      address6 = "2001:0DB8:02de:1000::0e13/64"
                      gateway6 = "2001:0DB8:02de:1000::1"
                      address = "10.158.180.45/24"
                      gateway = "10.158.180.1"
                      enabled = "0"

### CODE REVIEW

**Comment by xen-git** [comment]
Can one of the admins verify this patch for testing?

**Comment by Chunjie** [comment]
There are 4 pull requests to form a complete "vif-param-set  static-ip-setting" function, they are

1. https://github.com/xapi-project/xen-api/pull/2593, xapi change to add a new param "static-ip-setting" to vif-param-set, and propagate the change to xenopsd
2. https://github.com/xapi-project/xenopsd/pull/252, xenopsd change to write the "static-ip-setting" to xenstored
3. https://github.com/xapi-project/xcp-idl/pull/110, add an interface between xapi and xenopsd 
4. https://github.com/xapi-project/xenops-cli/pull/9, we add a field "static-ip-setting" to vif, this causes change to vif create interface

**Comment by Chunjie** [comment]
Travis CI outputs lots of the following message, not sure what has happened, looks like no error.

- Warning 50: ambiguous documentation comment
- File "db_actions.ml", line 1015, characters 2-9:
- Warning 50: ambiguous documentation comment
- File "db_actions.ml", line 1022, characters 2-9:
- Warning 50: ambiguous documentation comment
- File "db_actions.ml", line 1025, characters 2-9:
- Warning 50: ambiguous documentation comment
- File "db_actions.ml", line 1028, characters 2-9:
- Warning 50: ambiguous documentation comment
- File "db_actions.ml", line 1031, characters 2-9:
- Warning 50: ambiguous documentation comment
- File "db_actions.ml", line 1034, characters 2-9:
- Warning 50: ambiguous documentation comment
- File "db_actions.ml", line 1043, characters 2-9:
- Warning 50: ambiguous documentation comment
- File "db_actions.ml", line 1046, characters 2-9:

**Comment by robhoes** [comment]
@Chunjie, thanks for your patches. This looks like it is in pretty good shape already. However, I would like to see some changes to the API. Rather than a single string-string map field `static_ip_setting`, I prefer to have the individual settings exposed as first-class fields. This way we'd be better protected against missing or miss-spelled keys. I think this can be very similar to the way IP config is set on PIFs (for dom0).

The new VIF fields could be as follows (all `DynamicRO`):
```
(ip_configuration_mode) ipv4_configuration_mode
(string set) ipv4
(string) ipv4_gateway
(ip_configuration_mode) ipv6_configuration_mode
(string set) ipv6
(string) ipv6_gateway
```

...where `ip_configuration_mode` is a new enum that can be defined as:
```
let vif_ip_configuration_mode = Enum ("ip_configuration_mode",
	[ "None", "Follow the default IP configuration of the guest (this is guest-dependant)";
	  "Static", "Static IP address configuration" ])
```

Even if we would only support a single IPv4 address for the time being, it's worth making `ipv4` a set type so it is easier to extend in future.

The IP configuration mode replaces your `enabled` key. It also makes it easier to extend, for example if we'd want to request config from DHCP in future.

We can then add the following functions (similar to `PIF.reconfigure_ip`):
```
configure_ipv4(ip_configuration_mode mode, string set addresses, string gateway)
configure_ipv6(ip_configuration_mode mode, string set addresses, string gateway)
```
(where an "address" is a "address/prefixlen" string)

Next, we need a new feature flag from the guest agent, so that we can raise an error if the agent does not support this new mechanism yet (there will be lots of VMs with old tools). This is more reliable than checking `network_paths_optimized`.

It would also be good to add a xapi feature flag (ocaml/xapi/features.ml) and check it using `Pool_features.is_enabled`, so that the feature as a whole can be switched on and off.

Finally, we need to keep an eye on the `VIF.locking_mode` and ensure that the new API is compatible with it. If a VIF is locked to a particular IP address (`VIF.ip{v4,v6}_allowed`), we should raise an error if the IP that is being set does not match.

File: ocaml/xapi/xapi_vif.ml [hunk 1]
```diff
@@ -116,3 +162,23 @@ let add_ipv6_allowed ~__context ~self ~value =
 let remove_ipv6_allowed ~__context ~self ~value =
 	change_locking_config ~__context ~self ~licence_check:false
 		(fun () -> Db.VIF.remove_ipv6_allowed ~__context ~self ~value)
+
+let add_to_static_ip_setting ~__context ~self ~key ~value =
+	change_locking_config ~__context ~self ~licence_check:false
```
Comment (robhoes) [inline ocaml/xapi/xapi_vif.ml]: This should _not_ be `change_locking_config`, which for the VIF port locking feature.

File: ocaml/xapi/xapi_vif.ml [hunk 2]
```diff
@@ -116,3 +162,23 @@ let add_ipv6_allowed ~__context ~self ~value =
...
+		(fun () -> 
+			assert_ip_setting_is "static_ip_setting" key value;
+			let vm = Db.VIF.get_VM ~__context ~self in
+			let vm_gm = Db.VM.get_guest_metrics ~__context ~self:vm in 
+			let network_optimized = try Db.VM_guest_metrics.get_network_paths_optimized ~__context ~self:vm_gm with _ -> false in
```
Comment (robhoes) [inline ocaml/xapi/xapi_vif.ml]: This is where we should instead check a feature flag from the guest agent to assert that the agent support setting IP addresses.

**Comment by Chunjie** [comment]
"It would also be good to add a xapi feature flag (ocaml/xapi/features.ml) and check it using Pool_features.is_enabled, so that the feature as a whole can be switched on and off."

Hi @robhoes,

See REQ-201, there is no licensing requirement on this feature.

**Comment by stephen-turner** [comment]
@Chunjie, even if there's no licensing requirement now, we usually add licensing flags for new features, because we often get a request to license them at the last minute.

**Comment by Chunjie** [comment]
Hi @robhoes , I updated the patch.

File: ocaml/client_records/records.ml [hunk 3]
```diff
@@ -238,7 +238,7 @@ let pif_record rpc session_id pif =
       [
 	make_field ~name:"uuid"         ~get:(fun () -> (x ()).API.pIF_uuid) ();
 	make_field ~name:"device"       ~get:(fun () -> (x ()).API.pIF_device) ();
-	make_field ~name:"MAC"          ~get:(fun () -> (x ()).API.pIF_MAC) 
+	make_field ~name:"MAC"          ~get:(fun () -> (x ()).API.pIF_MAC)
```
Comment (robhoes) [inline ocaml/client_records/records.ml]: Unnecessary diff here.

File: ocaml/xapi/cancel_tests.ml [hunk 4]
```diff
@@ -105,7 +105,7 @@ let find_or_create_vif { session_id = session_id; vm = vm; net = net } =
 		let vif, _ = List.find (fun (_, r) -> r.API.vIF_device = vif_idx) (List.combine vifs vif_records) in
 		vif
 	with Not_found ->
-		Client.VIF.create ~rpc ~session_id ~vM:vm ~network:net ~mAC:"" ~device:vif_idx ~mTU:1500L ~other_config:[] ~qos_algorithm_type:"" ~qos_algorithm_params:[] ~locking_mode:`network_default ~ipv4_allowed:[] ~ipv6_allowed:[] 
+		Client.VIF.create ~rpc ~session_id ~vM:vm ~network:net ~mAC:"" ~device:vif_idx ~mTU:1500L ~other_config:[] ~qos_algorithm_type:"" ~qos_algorithm_params:[] ~locking_mode:`network_default ~ipv4_allowed:[] ~ipv6_allowed:[]
```
Comment (robhoes) [inline ocaml/xapi/cancel_tests.ml]: Unnecessary diff here.

File: ocaml/xapi/cli_operations.ml [hunk 5]
```diff
@@ -1574,6 +1574,28 @@ let vif_unplug printer rpc session_id params =
 	let force = get_bool_param params "force" in
 	(if force then Client.VIF.unplug_force else Client.VIF.unplug) rpc session_id vif
 
+let vif_configure_ipv4 printer rpc session_id params =
+	let read_optional_case_insensitive key =
```
Comment (robhoes) [inline ocaml/xapi/cli_operations.ml]: I don't think we need this function here (and below). The `pif_reconfigure_ip` function has one, because it wants to accept both `IP` and `ip`, and `DNS` and `dns` as arguments. That doesn't really apply here.

File: ocaml/xapi/cli_operations_geneva.ml [hunk 6]
```diff
@@ -368,7 +368,7 @@ let vm_install fd printer rpc session_id params =
 	let add_vif net =
 	  let mac = Record_util.random_mac_local () in
 	  marshal fd (Command (Print ("Adding VIF, device "^(string_of_int !device)^" to network '"^(Client.Network.get_name_label rpc session_id net)^"' mac="^mac)));
-	  ignore(Client.VIF.create rpc session_id (string_of_int !device) net new_vm mac 1500L [] "" [] `network_default [] [] );
```
Comment (robhoes) [inline ocaml/xapi/cli_operations_geneva.ml]: Unnecessary diff.

File: ocaml/xapi/message_forwarding.ml [hunk 7]
```diff
@@ -2665,7 +2665,7 @@ module Forward = functor(Local: Custom_actions.CUSTOM_ACTIONS) -> struct
 
 		(* -------------------------------------------------------------------------- *)
 
-		let create ~__context ~device ~network ~vM ~mAC ~mTU ~other_config ~qos_algorithm_type ~qos_algorithm_params =
+		let create ~__context ~device ~network ~vM ~mAC ~mTU ~other_config ~qos_algorithm_type ~qos_algorithm_params  =
```
Comment (robhoes) [inline ocaml/xapi/message_forwarding.ml]: Unnecessary diff.

File: ocaml/xapi/message_forwarding.ml [hunk 8]
```diff
@@ -2737,6 +2737,21 @@ module Forward = functor(Local: Custom_actions.CUSTOM_ACTIONS) -> struct
...
+		let configure_ipv4 ~__context ~self ~mode ~address ~gateway =
+			info "VIF.configure_ipv4: VIF = '%s'; mode = '%s'; address = '%s'; gateway = '%s'"
+				(vif_uuid ~__context self)
+				(Record_util.vif_ipv4_configuration_mode_to_string mode) address gateway;
+			let local_fn = Local.VIF.configure_ipv4 ~self ~mode ~address ~gateway in
+			let remote_fn = (fun session_id rpc -> Client.VIF.configure_ipv4 rpc session_id self mode address gateway) in
+			forward_vif_op ~local_fn ~__context ~self remote_fn
+
+		let configure_ipv6 ~__context ~self ~mode ~address ~gateway =
+			info "PIF.configure_ipv6: PIF = '%s'; mode = '%s'; address = '%s'; gateway = '%s'"
```
Comment (robhoes) [inline ocaml/xapi/message_forwarding.ml]: `PIF` -> `VIF`.

File: ocaml/xapi/message_forwarding.ml [hunk 9]
```diff
@@ -2737,6 +2737,21 @@ module Forward = functor(Local: Custom_actions.CUSTOM_ACTIONS) -> struct
...
+				(vif_uuid ~__context self)
+				(Record_util.ipv6_configuration_mode_to_string mode) address gateway;
```
Comment (robhoes) [inline ocaml/xapi/message_forwarding.ml]: Should be `vif_ipv6_configuration_mode_to_string`.

File: ocaml/xapi/quicktest.ml [hunk 10]
```diff
@@ -779,7 +779,7 @@ let async_test session_id =
     | _ -> failwith "Expecting 1 new disk!"
 
 let make_vif ~session_id ~vM ~network ~device = 
-  Client.VIF.create ~rpc:!rpc ~session_id ~vM ~network ~mTU:0L ~mAC:"" ~device ~other_config:["promiscuous", "on"; "mtu", "1400"] ~qos_algorithm_type:"" ~qos_algorithm_params:[] 
+  Client.VIF.create ~rpc:!rpc ~session_id ~vM ~network ~mTU:0L ~mAC:"" ~device ~other_config:["promiscuous", "on"; "mtu", "1400"] ~qos_algorithm_type:"" ~qos_algorithm_params:[]
```
Comment (robhoes) [inline ocaml/xapi/quicktest.ml]: Unnecessary diff.

File: ocaml/xapi/xapi_vif.ml [hunk 11]
```diff
@@ -22,6 +23,36 @@ let assert_operation_valid ~__context ~self ~(op:API.vif_operations) =
 let update_allowed_operations ~__context ~self : unit =
   update_allowed_operations ~__context ~self
 
+let assert_license_feature_enabled ~__context =
```
Comment (robhoes) [inline ocaml/xapi/xapi_vif.ml]: There is already a generic function for this: `Pool_features.assert_enabled`.

File: ocaml/xapi/xapi_vif.ml [hunk 12]
```diff
@@ -116,3 +192,103 @@ let add_ipv6_allowed ~__context ~self ~value =
 let remove_ipv6_allowed ~__context ~self ~value =
 	change_locking_config ~__context ~self ~licence_check:false
 		(fun () -> Db.VIF.remove_ipv6_allowed ~__context ~self ~value)
+
+let configure_ipv4 ~__context ~self ~mode ~address ~gateway =
+	assert_license_feature_enabled ~__context;
```
Comment (robhoes) [inline ocaml/xapi/xapi_vif.ml]: Use `Pool_features.assert_enabled` here.

**Comment by robhoes** [comment]
The functions in xapi_vif.ml are a bit messy. We need to de-duplicate the code a bit. In fact, there are lots of "check this IP address" functions scattered around the code base, which we should factor out. I'll write a patch for this now.

**Comment by Chunjie** [comment]
obsolete patch, a new pull-request is on the way
